dace/codegen/targets/cpu.py

```diff
@@ -277,16 +283,26 @@ def allocate_array(self, sdfg, dfg, state_id, node, nodedesc, function_stream, d
         # Check if array is already declared
         declared = self._dispatcher.declared_arrays.has(alloc_name)
 
+        define_var = self._dispatcher.defined_vars.add
+        # NOTE: The code below fixes symbol-related issues with transient data originally defined in a NestedSDFG scope
+        # but promoted to be persistent. These data must have their free symbols replaced with the corresponding
+        # top-level SDFG symbols.
+        if nodedesc.lifetime == dtypes.AllocationLifetime.Persistent:
+            define_var = self._dispatcher.defined_vars.add_global
+            if sdfg.parent and any(str(s) in sdfg.parent_nsdfg_node.symbol_mapping for s in nodedesc.free_symbols):
```

Comment:
what if it's two layers down and in the grandparent SDFG it is not defined as a free symbol (but e.g., set in an interstate edge)

Reply:
what I mean is that the while loop below should cover that and fail gracefully